src/StreamJsonRpc/NerdbankMessagePackFormatter.cs

```diff
@@ -0,0 +1,2109 @@
+                }
+            }
+
+            public override void Write(ref MessagePackWriter writer, in T? value, SerializationContext context)
+            {
+                context.DepthStep();
+
+                if (value is null)
+                {
+                    writer.WriteNil();
+                    return;
+                }
+
+                exceptionRecursionCounter.Value++;
+                try
+                {
+                    if (exceptionRecursionCounter.Value > formatter.JsonRpc?.ExceptionOptions.RecursionLimit)
+                    {
+                        // Exception recursion has gone too deep. Skip this value and write null as if there were no inner exception.
+                        writer.WriteNil();
+                        return;
+                    }
+
+                    // TODO: Is this the right context?
+                    var info = new SerializationInfo(typeof(T), new MessagePackFormatterConverter(formatter.rpcProfile));
+                    ExceptionSerializationHelpers.Serialize(value, info);
+                    writer.WriteMapHeader(info.GetSafeMemberCount());
+                    foreach (SerializationEntry element in info.GetSafeMembers())
+                    {
+                        writer.Write(element.Name);
+                        formatter.rpcProfile.SerializeObject(
+                            ref writer,
+                            element.Value,
+                            element.ObjectType,
+                            context.CancellationToken);
+                    }
+                }
+                finally
+                {
+                    exceptionRecursionCounter.Value--;
+                }
+            }
+
+            public override JsonObject? GetJsonSchema(JsonSchemaContext context, ITypeShape typeShape)
+            {
+                return CreateUndocumentedSchema(typeof(ExceptionConverter<T>));
+            }
+        }
+    }
+
+    [GenerateShape<JsonRpcMessage>]
+    private partial class JsonRpcMessageConverter : MessagePackConverter<JsonRpcMessage>
+    {
+        private readonly NerdbankMessagePackFormatter formatter;
+
+        internal JsonRpcMessageConverter(NerdbankMessagePackFormatter formatter)
+        {
+            this.formatter = formatter;
+        }
+
+        public override JsonRpcMessage? Read(ref MessagePackReader reader, SerializationContext context)
+        {
+            context.DepthStep();
+
+            MessagePackReader readAhead = reader.CreatePeekReader();
+            int propertyCount = readAhead.ReadMapHeader();
+            for (int i = 0; i < propertyCount; i++)
+            {
+                if (MethodPropertyName.TryRead(ref readAhead))
+                {
+                    return context.GetConverter<Protocol.JsonRpcRequest>(context.TypeShapeProvider).Read(ref reader, context);
+                }
+                else if (ResultPropertyName.TryRead(ref readAhead))
+                {
```

Comment:
@AArnott - After adopting MessagePackString, this converter always fails throwing the UnrecognizedJsonRpcMessageException exception at the end of the method.

Reply:
I can WriteRaw(MyMPString.MsgPck.Span) and then ReadString() just fine.